src/cocotb/share/lib/embed/gpi_embed.cpp

```diff
@@ -135,20 +117,89 @@ extern "C" COCOTB_EXPORT void _embed_init_python(void) {
+            // LCOV_EXCL_START
             LOG_ERROR("Invalid log level: %s", log_level);
+            // LCOV_EXCL_STOP
         }
     }
 
-    to_python();
     // must set program name to Python executable before initialization, so
     // initialization can determine path from executable
-    set_program_name_in_venv();
-    Py_Initialize(); /* Initialize the interpreter */
+
+    static wchar_t interpreter_path[PATH_MAX], sys_executable[PATH_MAX];
+
+    if (get_interpreter_path(interpreter_path, sizeof(interpreter_path))) {
+        // LCOV_EXCL_START
+        return;
+        // LCOV_EXCL_STOP
+    }
+    LOG_INFO("Using Python interpreter at %ls", interpreter_path);
+
+#if PY_VERSION_HEX >= 0x3080000
+    /* Use the new Python Initialization Configuration from Python 3.8. */
+    PyConfig config;
+    PyStatus status;
+
+    PyConfig_InitPythonConfig(&config);
+    DEFER(PyConfig_Clear(&config));
+
+    PyConfig_SetString(&config, &config.program_name, interpreter_path);
+
+    status = PyConfig_SetArgv(&config, 1, argv);
+    if (PyStatus_Exception(status)) {
+        // LCOV_EXCL_START
+        LOG_ERROR("Failed to set ARGV during the Python initialization");
+        if (status.err_msg != NULL) {
+            LOG_ERROR("\terror: %s", status.err_msg);
+        }
+        if (status.func != NULL) {
+            LOG_ERROR("\tfunction: %s", status.func);
+        }
+        return;
+        // LCOV_EXCL_STOP
+    }
+
+    status = Py_InitializeFromConfig(&config);
+    if (PyStatus_Exception(status)) {
+        // LCOV_EXCL_START
+        LOG_ERROR("Failed to initialize Python");
+        if (status.err_msg != NULL) {
+            LOG_ERROR("\terror: %s", status.err_msg);
+        }
+        if (status.func != NULL) {
+            LOG_ERROR("\tfunction: %s", status.func);
+        }
+        return;
+        // LCOV_EXCL_STOP
+    }
+#else
+    /* Use the old API. */
+    Py_SetProgramName(interpreter_path);
+    Py_Initialize();
     PySys_SetArgvEx(1, argv, 0);
+#endif
 
-    /* Swap out and return current thread state and release the GIL */
-    gtstate = PyEval_SaveThread();
-    to_simulator();
+    /* Sanity check: make sure sys.executable was initialized to
+     * interpreter_path. */
+    PyObject *sys_executable_obj = PySys_GetObject("executable");
+    if (sys_executable_obj == NULL) {
+        // LCOV_EXCL_START
+        LOG_ERROR("Failed to load sys.executable");
+        return;
+        // LCOV_EXCL_STOP
+    } else if (PyUnicode_AsWideChar(sys_executable_obj, sys_executable,
+                                    sizeof(sys_executable)) == -1) {
+        // LCOV_EXCL_START
+        LOG_ERROR("Failed to convert sys.executable to wide string");
+        return;
+        // LCOV_EXCL_STOP
+    } else if (wcscmp(interpreter_path, sys_executable) != 0) {
+        // LCOV_EXCL_START
+        LOG_ERROR("Unexpected sys.executable value (expected '%ls', got '%ls')",
+                  sys_executable, interpreter_path);
```

Comment:
I mistakenly got this the other way round.

```suggestion
        LOG_ERROR("Unexpected sys.executable value (expected '%ls', got '%ls')",
                  interpreter_path, sys_executable);
```